contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,377 @@
+        // encoded as an int64. On Unix, the command `date +%s%N` returns
+        // the  current time in nanoseconds since 1970-01-01 00:00:00 UTC.
+        // This field is REQUIRED.
+	CreatedAt int64 `json:"createdAt,omitempty" protobuf:"varint,3,opt,name=createdAt"`
+}
+
+```
+
+#### The `DataSource` Object in PVC
+
+Add a new field into PVC to represent the source of the data which is prepopulated to the provisioned volume. Possible source types may include the following:
+
+    * VolumeSnapshot: restore snapshot to a new volume
+    * PersistentVolumeClaim: clone volume which is represented by PVC
+
+```
+
+type PersistentVolumeClaimSpec struct {
+        // If specified, volume will be prepopulated with data from the PVCDataSourceRef.
+        // +optional
+        PVCDataSourceRef *core_v1.LocalObjectReference `json:"dataSourceRef" protobuf:"bytes,2,opt,name=dataSourceRef"`
+
+}
+
+```
+
+The existing LocalObjectReference in core API will be modified to add a `Kind`.
+
+```
+
+// LocalObjectReference contains enough information to let you locate the referenced object inside the same namespace.
+type LocalObjectReference struct {
+        //TODO: Add other useful fields.  apiVersion, kind, uid?
+        Name string
+	// Kind indicates the type of the object reference.
+	// +optional
+	Kind string
+}
+
+```
+
+In the first version, only VolumeSnapshot will be a supported `Type` for data source object reference. PersistentVolumeClaim will be added in a future version. If unsupported `Type` is used, the PV Controller SHALL bail out of the operation.
+
+
+#### The `VolumeSnapshotClass` Object
+
+A new VolumeSnapshotClass API object will be added to avoid mixing parameters between snapshots and volumes. Each CSI Volume Driver can have its own default VolumeSnapshotClass. If VolumeSnapshotClass is not provided, a default will be used. It allows to add new parameters for snapshots.
+
+```
+
+type VolumeSnapshotClass struct {
+        metav1.TypeMeta `json:",inline"`
+        // +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Snapshotter is the driver expected to handle this VolumeSnapshotClass.
+	// This value may not be empty.
+	Snapshotter string
+
+	// Parameters holds parameters for the snapshotter.
+        // These values are opaque to the system and are passed directly
+        // to the snapshotter. The only validation done on keys is that they are
+        // not empty. The maximum number of parameters is
+        // 512, with a cumulative max size of 256K
+        // +optional
+	Parameters map[string]string
+}
+
+```
+
+#### Add `DataSource` to `StorageClass`
+
+DataSource will be added as an optional parameter in `StorageClass` to allow volumes belonging to the same `StorageClass` to be created with prepopulated data.
+
+```
+
+type StorageClass struct {
+        // If specified, volume will be prepouplated with data from the DataSource
```

Comment:
Adam on the call had a good suggestion of adding DataSource to PV as well as PVC because if an external-provisioner doesn't understand the new DataSource field and doesn't pre-populate the new volume, PV/PVC provisioner should be able to detect that (e.g. PVC has data source but PV does not)

Reply:
Done